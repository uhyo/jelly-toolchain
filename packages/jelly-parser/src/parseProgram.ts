import { parseChar } from "./parseChar";
import { ParserState } from "./parserState";

export const parseProgram = (program: string): void => {
  const state: ParserState = {
    lines: [],
    current: [],
    offset: 0,
  };

  for (let i = 0; i < program.length; i++) {
    parseChar(state, program[i]);
  }
  return;
};
