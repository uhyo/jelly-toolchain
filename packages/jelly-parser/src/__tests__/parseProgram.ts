import { parseProgram } from "../parseProgram";

describe("parseProgram", () => {
  it("ɠḲV×/ị“Odd“Even”", () => {
    expect(parseProgram("ɠḲV×/ị“Odd“Even”")).toMatchSnapshot();
  });
});
