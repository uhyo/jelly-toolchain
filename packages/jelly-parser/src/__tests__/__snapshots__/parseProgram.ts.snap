// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseProgram ɠċ”1 1`] = `
Object {
  "links": Array [
    Object {
      "links": Array [
        Object {
          "arity": 0,
          "name": "ɠ",
          "type": "atom",
        },
        Object {
          "arity": 2,
          "name": "ċ",
          "type": "atom",
        },
        Object {
          "arity": 0,
          "type": "charLiteral",
          "value": "1",
        },
      ],
      "type": "chain",
    },
  ],
}
`;

exports[`parseProgram ɠḲV×/ị“Odd“Even” 1`] = `
Object {
  "links": Array [
    Object {
      "links": Array [
        Object {
          "arity": 0,
          "name": "ɠ",
          "type": "atom",
        },
        Object {
          "arity": 1,
          "name": "Ḳ",
          "type": "atom",
        },
        Object {
          "arity": 1,
          "name": "V",
          "type": "atom",
        },
        Object {
          "arity": 2,
          "name": "×",
          "type": "atom",
        },
        Object {
          "arity": 1,
          "link": Object {
            "arity": 2,
            "name": "×",
            "type": "atom",
          },
          "quick": "reduce",
          "type": "quick",
        },
        Object {
          "arity": 2,
          "name": "ị",
          "type": "atom",
        },
        Object {
          "arity": 0,
          "array": true,
          "type": "stringLiteral",
          "values": Array [
            "Odd",
            "Even",
          ],
        },
      ],
      "type": "chain",
    },
  ],
}
`;
